### PR TITLE
fix: use SelectRequestBuilder for rpc call

### LINKED ISSUE #200: RPC call expects `data` in APIResponse to be a list, which is not mandatory
# Bug report

## Describe the bug

The response of an RPC query expects `data` as `List[Dict[str, Any]]` in the `APIResponse` object found in `postgrest/base_request_builder.py`, but it must not be a list.

This call worked in version `0.10.3`, but does not in version `0.10.4`

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Install this function in the PostgreSQL database ([from here](https://github.com/supabase-community/supabase-custom-claims)) to get the user's `raw_app_meta_data` from the Supabase user object:

```sql
CREATE OR REPLACE FUNCTION get_claims(uid uuid) RETURNS "jsonb"
    LANGUAGE "plpgsql" SECURITY DEFINER SET search_path = public
    AS $$
    DECLARE retval jsonb;
    BEGIN
      IF NOT is_claims_admin() THEN
          RETURN '{"error":"access denied"}'::jsonb;
      ELSE
        select raw_app_meta_data from auth.users into retval where id = uid::uuid;
        return retval;
      END IF;
    END;
$$;
```
3. Make the respective RPC call with an authorized Supabase client

```python
sb.rpc(
        fn="get_claims",
        params={"uid": str(user_id)}
    ).execute()
```

## Expected behavior

I expect to receive the response body _as is_. 

## System information

- OS: macOs
- Version of postgrest-py: 0.10.4

## Additional context

Thrown error: 

```shell
Traceback (most recent call last):
  File "/Users/timreibe/VENV/supabase-stripe-webhook/lib/python3.9/site-packages/postgrest/_sync/request_builder.py", line 65, in execute
    return APIResponse.from_http_request_response(r)
  File "/Users/timreibe/VENV/supabase-stripe-webhook/lib/python3.9/site-packages/postgrest/base_request_builder.py", line 157, in from_http_request_response
    return cls(data=data, count=count)
  File "pydantic/main.py", line 342, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for APIResponse
data
  value is not a valid list (type=type_error.list)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/timreibe/Documents/Code/py/supabase-stripe-webhook/app.py", line 29, in <module>
    user_id = _get_custom_claims("f25f0ca7-4116-4587-a445-084057d52b50")
  File "/Users/timreibe/Documents/Code/py/supabase-stripe-webhook/src/sb/custom_claims.py", line 64, in _get_custom_claims
    print(res.execute())
  File "/Users/timreibe/VENV/supabase-stripe-webhook/lib/python3.9/site-packages/postgrest/_sync/request_builder.py", line 69, in execute
    raise APIError(r.json()) from e
postgrest.exceptions.APIError: {'scid': 'STRIPE_USER_ID', 'provider': 'email', 'providers': ['email']}
```

### PR DESCRIPTION
**Pull request by a5r0n**
fixes #200

### CODE REVIEW

**Comment by a5r0n** [comment]
@J0 @anand2312 can you take a look?

File: postgrest/_async/client.py [hunk 1]
```diff
@@ -76,24 +76,24 @@ def from_table(self, table: str) -> AsyncRequestBuilder:
         """Alias to :meth:`from_`."""
         return self.from_(table)
 
-    async def rpc(self, func: str, params: dict) -> AsyncFilterRequestBuilder:
+    async def rpc(self, func: str, params: dict) -> AsyncSelectRequestBuilder:
```
Comment (a5r0n) [inline postgrest/_async/client.py:79]: is there any reason to use async for this function?
`await (await client.rpc(...)).execute()`

```suggestion
    def rpc(self, func: str, params: dict) -> AsyncSelectRequestBuilder:
```

**Comment by anand2312** [comment]
Sorry for not getting to this earlier
I'm leaning towards not having a pydantic model as the response of .rpc() here, and just return the JSON - feels like the most intuitive solution to me that covers every case. What do you guys think?

**Comment by a5r0n** [comment]
from response perspective, there is 3 result types.

* when the function returns `setof` records, (e.g. `CREATE  FUNCTION find_users(...)
  RETURNS SETOF "User"`) 
  for that, the current implementation works well.
 * my use case - when the function returns object (e.g. `CREATE  FUNCTION get_user_id(..., OUT id integer)`)
 * the last one, is when the function returns simple value, like what @chrisgoddard is looking for, (`CREATE  FUNCTION get_user_id(...) RETURNS integer`)

**Comment by anand2312** [comment]
And I think functions can return None as well with `RETURNS void` iirc @a5r0n

**Comment by chrisgoddard** [comment]
My 2c for right now would be to at least provide the mechanism for returning the raw response (not the raw HTTP response - we can still call `.json()`) and then let users define how they want to use the response - I feel like there could be too many different edge cases of how people are using stored procedures within Postgres.

Right now I'm doing this as a workaround:
```python

def rpc_call(
    supabase: SupabaseClient,
    procedure: str,
    params: dict = {},
    type_: Optional[Type[BaseModel]] = None
):
    query = supabase.postgrest.rpc(
        procedure,
        orjson.loads(orjson.dumps(params))  # ensures json properly serialized for requests
    )
    data = query.session.request(
        query.http_method,
        query.path,
        json=query.json,
        params=query.params,
        headers=query.headers,
    ).json()

    if data:
        if type_ and isinstance(data, dict):
            return type_(**data)
        return data
```

Ultimately I could imagine a more sophisticated solution where you could define a remote procedure explicitly in your client (could use Pydantic types) where the inputs to the function could also be validated as well as the response types - but for now I think just making it so people can use stored procedures however they have them configured would be the simplest.

**Comment by a5r0n** [comment]
@chrisgoddard the official supabase client is actually using a different library to invoke functions, did you try that? (https://github.com/supabase-community/functions-py)

**Comment by chrisgoddard** [comment]
> @chrisgoddard the official supabase client is actually using a different library to invoke functions, did you try that? (https://github.com/supabase-community/functions-py)

Those are for edge functions - I'm talking about Postgres stored procedures.

**Comment by a5r0n** [comment]
Sure, my mistake.

**Comment by a5r0n** [comment]
@anand2312 
can we get this PR merged?
at least for my use case, this solution works fine.

**Comment by anand2312** [comment]
I'm reluctant to merge because it doesn't really fix the issue at hand 😦 If I'm not able to come up with something better in the next day or two I guess I'll merge this

**Comment by ShantanuNair** [comment]
@anand2312 @a5r0n Is there any update on this? Appreciate you guys working on this. I just built a bunch of stuff around the assumption that I Can use RPC calls and only now found out that this is broken on supabase-py/postgrest-py

**Comment by anand2312** [comment]
#309 was merged instead of this as it was a more complete fix - but thanks for the contribution regardless! @a5r0n

**Pull request closed without merging**